reactor-extra/src/main/java/reactor/retry/RetryBuilder.java

```diff
@@ -0,0 +1,440 @@
+		long firstBackoffMs = firstBackoff.toMillis();
+		long maxBackoffMs = maxBackoff != null ? maxBackoff.toMillis() : Long.MAX_VALUE;
+		if (maxBackoffMs < firstBackoffMs)
+			throw new IllegalArgumentException("maxBackoff must be > firstBackoff");
+		AtomicLong prevBackoffMs = new AtomicLong();
+		ThreadLocalRandom random = ThreadLocalRandom.current();
+		this.backoffCalculator = iteration -> {
+			long maxMs = Math.max(firstBackoffMs, prevBackoffMs.get() * 3);
+			long jitterBackoffMs = firstBackoffMs == maxMs ? firstBackoffMs : random.nextLong(firstBackoffMs, maxMs);
+			prevBackoffMs.set(jitterBackoffMs);
+			return Duration.ofMillis(Math.min(maxBackoffMs, jitterBackoffMs));
+		};
+		return this;
+	}
+
+	/**
+	 * Configures a scheduler for backoff delays.
+	 * @param scheduler a scheduler instance
+	 * @return updated {@link RetryBuilder}
+	 */
+	public RetryBuilder backoffScheduler(Scheduler scheduler) {
+		this.backoffScheduler = scheduler;
+		return this;
+	}
+
+	/**
+	 * Configures a retry context which may include application context. Applications which need access to
+	 * application state for rollbacks may set a retry context for easy access to the state in
+	 * {@link #doOnRetry(Consumer)} callbacks.
+	 *
+	 * @param retryContext RetryContext instance which is included in {@link #doOnRetry(Consumer)} callbacks
+	 * @return updated {@link RetryBuilder}
+	 */
+	public RetryBuilder retryContext(RetryContext<?> retryContext) {
+		this.retryContext = retryContext;
+		return this;
+	}
+
+	/**
+	 * Configures an <code>onRetry</code> callback that is invoked prior to every retry
+	 * attempt, before any backoff delay. Any rollbacks that need to be performed before
+	 * retry can be performed in the callback. Application context required for rollback
+	 * may be configured using {@link #retryContext(RetryContext)}.
+	 *
+	 * @param onRetry callback for retry notification
+	 * @return updated {@link RetryBuilder}
+	 */
+	public RetryBuilder doOnRetry(Consumer<RetryContext<?>> onRetry) {
+		this.onRetry = onRetry;
+		return this;
+	}
+
+	/**
+	 * Returns a function that may be used to control retries when used with:
+	 * <ul>
+	 *   <li>{@link Flux#retryWhen(Function)}</li>
+	 *   <li>{@link Mono#retryWhen(Function)}</li>
+	 * </ul>
+	 * If a retry is not attempted because the exception is not retriable or the
+	 * custom retry predicate returns false, the original exception from the last
+	 * attempt is propagated as the error. If a retry is not attempted because
+	 * retries have been exhausted due to timeout or maximum attempt limit,
+	 * a {@link RetryExhaustedException} is generated with the original exception
+	 * from the last attempt as the cause.
+	 *
+	 * @return {@link Function} providing a {@link Flux} signalling any error from the
+	 *         source sequence and returning a {@link Publisher} companion.
+	 */
+	public Function<Flux<Throwable>, ? extends Publisher<?>> buildRetry() {
+		AtomicLong iteration = new AtomicLong();
```

Comment:
I think we could use a `zipWith(Flux.range(1, MAX_VALUE))` here instead of incrementing an `AtomicLong`. wdyt @smaldini ?